bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/TeaserImpl.java

```diff
@@ -0,0 +1,193 @@
+
+    @ScriptVariable
+    protected Page currentPage;
+
+    @Inject
+    private SlingSettingsService settings;
+
+    @PostConstruct
+    protected void initModel() {
+        compHashCode = properties.get(Teaser.PN_TRACKING_OBJECT_ID, compHashCode);
+        trackingEnabled = properties.get(Teaser.PN_TRACKING_ENABLED, trackingEnabled);
+        super.initModel();
+        hashCodeValue = String.valueOf(Math.abs(resource.getPath().hashCode() - 1));
+        runmode = settings.getRunModes();
+        if (runmode.contains(Externalizer.AUTHOR)) {
+            populateObjectId();
+        }
+    }
+
+    @Override
+    protected void populateActions() {
+        Resource actionsNode = resource.getChild(Teaser.NN_ACTIONS);
+        if (actionsNode != null) {
+            for (Resource action : actionsNode.getChildren()) {
+                actions.add(new ListItem() {
+
+                    private ValueMap properties = action.getValueMap();
+                    private String title = properties.get(PN_ACTION_TEXT, String.class);
+                    private String url = properties.get(PN_ACTION_LINK, String.class);
+                    private Page page = null;
+                    {
+                        if (url != null && url.startsWith("/")) {
+                            page = pageManager.getPage(url);
+                        }
+                    }
+
+                    @Nullable
+                    @Override
+                    public String getTitle() {
+                        return title;
+                    }
+
+                    @Nullable
+                    @Override
+                    @JsonIgnore
+                    public String getPath() {
+                        return url;
+                    }
+
+                    @Nullable
+                    @Override
+                    public String getURL() {
+                        if (page != null) {
+                            return Utils.getURL(request, page);
+                        } else {
+                            return url;
+                        }
+                    }
+
+                    @Nullable
+                    @Override
+                    public String getLinkTrackingCode() {
+                        linkTrackingCode = new StringBuilder();
+                        counter++;
+                        if (trackingEnabled && !compHashCode.isEmpty()) {
+                            linkTrackingCode = linkTrackingCode.append(OBJECT_ID).append(APOSTROPHE)
+                                    .append(compHashCode).append(StringUtils.SPACE).append(counter).append(APOSTROPHE)
+                                    .append(SEMICOLON);
+                        }
+                        return linkTrackingCode.toString();
+                    }
+                });
+            }
+        }
+    }
+
+    @Override
+    public List<ListItem> getActions() {
+        return actions;
+    }
+
+    @Override
+    public String getLinkTrackingCode() {
```

Comment:
Looks like a lot of code dupe between models - could this be abstracted out to a "helper" LinkTracking service/model?